Issue #21 - create a new enum type and use it to query the general result, which replaces methods using constructor as an input.

alastor0325 opened the pull request.

alastor0325 commented:
Not sure why I can't set @tjenkinson as my reviewer manually, so I would tag instead to see if you have anytime to help me review this change. Thanks!

tjenkinson reviewed (approved):
So it’s an enum, or instance? LGTM

chrisn commented on index.bs:
```diff
@@ -127,60 +129,94 @@ div.enum-description > table > thead > tr > th {
     <pre class="idl">
-      callback HTMLMediaElementConstructor = HTMLMediaElement ();
-      callback AudioContextConstructor = AudioContext ();
+      enum AutoplayPolicyMediaType { "mediaElement", "audioContext" };
```
The WebIDL spec [recommends](https://webidl.spec.whatwg.org/#idl-enums) using all lower case for enumeration values, unless there's specific reason no to, and to consider related uses to decide whether to use dashes to separate word.

Looking at media related specs, [Media Capabilities](https://www.w3.org/TR/media-capabilities/#mediadecodingtype) uses dashes but [Media Session](https://www.w3.org/TR/mediasession/#the-mediasession-interface) does not.

Here, we're using an enum to indicate an interface. I found an example in the HTML spec: `arraybuffer` in the [WebSocket interface](https://html.spec.whatwg.org/multipage/web-sockets.html#the-websocket-interface), so I suggest following that and using lower case without dashes here, so `mediaelement` and `audiocontext`.

tidoust replied:
FWIW, an up-to-date list of Web IDL enum values defined across specs can be found in [WebIDLpedia](https://dontcallmedom.github.io/webidlpedia/enum.html). I could not spot other examples of enum values that reference an IDL type or interface.

chrisn commented on index.bs:
```diff
@@ -127,60 +129,94 @@ div.enum-description > table > thead > tr > th {
+      the following steps:
+      <ol>
+        <li>
+          If {{type}} is not {{AutoplayPolicyMediaType/mediaElement}} or
```
Do we need to specify this step? AFAIK use of WebIDL enum requires this behaviour already.

chrisn commented on index.bs:
```diff
@@ -127,60 +129,94 @@ div.enum-description > table > thead > tr > th {
+      <dt>If the return value is {{allowed}}</dt>
       <dd>
-        All media are allowed to autoplay.
+        All media, corresponding with given type, are allowed to autoplay.
```
Minor nit: "with the given type"

chrisn commented on index.bs:
```diff
@@ -61,7 +61,7 @@ div.enum-description > table > thead > tr > th {
   Autoplay detection can be performed through the {{Navigator}} object. The
   result can either allow authors to know if media, which have the same type of
-  the queried constructor and exist in the {{document}} contained in the
+  the queried media type and exist in the {{document}} contained in the
```
Minor nit: I suggest using "given" rather than "queried"

chrisn commented on index.bs:
```diff
@@ -127,60 +129,94 @@ div.enum-description > table > thead > tr > th {
+    </div>
+
+    <h4 id="query-by-a-media-type">Query by a Media Type</h4>
+      The {{getAutoplayPolicy(type)}} methods return the **rough** status of
```
Can we be more precise than "rough"? If it's true that the method returns the correct policy at the time the method is called, but the policy may change, can we set some expectation for when the value is stable, eg., in terms of the event loop?

alastor0325 replied:
The detailed explanation about this are in the following examples and notes, `rough` means it's not 100% correct because the query by an element can return a different result. "Rough" isn't indicating the characteristic of the result it time various, because all of those results returned by different API are having that characteristic. 

But yes, I will add more explanation for this part. Thanks!

chrisn reviewed (changes requested):
Thanks @alastor0325, this is looking good. I added a few comments to consider.

alastor0325 commented:
Hmm sorry I think I messed up this PR with an incorrect commit which contains some merging conflicts. I'm trying to find a way to handle the conflicts correctly and make the PR only containning the correct change....

alastor0325 commented:
Updated PR and addressed review suggestions, this time the changes in PR seems correct. Hopefully what I was doing for handling the conflicts is correct.

chrisn commented on index.bs:
```diff
@@ -132,78 +134,109 @@ div.enum-description > table > thead > tr > th {
+        AutoplayPolicy getAutoplayPolicy(AutoplayPolicyMediaType type);
         AutoplayPolicy getAutoplayPolicy(HTMLMediaElement element);
-        AutoplayPolicy getAutoplayPolicy(AudioContext context);
+        AutoplayPolicy getAutoplayPolicy(audiocontext context);
```
This one should still be `AudioContext`

chrisn commented on index.bs:
```diff
@@ -96,8 +96,10 @@ div.enum-description > table > thead > tr > th {
             <div class="note">
-              Note: Currently, this attribute will only be returned when a queried
-              object or a constructor is a type of {{HTMLMediaElement}}.
+              Note: Currently, this attribute will only be returned when a
```
I suggest changing "a given" to "the given"

chrisn commented on index.bs:
```diff
@@ -132,78 +134,109 @@ div.enum-description > table > thead > tr > th {
+        users.
+
+        For example, at first, the result of querying by a media type and
+        querying by a object would both be {{disallowed}}. After a user clicks
```
Suggest changing "a object" to "an object"

chrisn commented on index.bs:
```diff
@@ -132,78 +134,109 @@ div.enum-description > table > thead > tr > th {
+      </div>
+
+      <div class=example>
+        Some user-agents may not allow any media element to autoplay by default,
```
For consistency, I suggest writing "user agent" without hyphen throughout

chrisn commented on index.bs:
```diff
@@ -267,7 +301,7 @@ div.enum-description > table > thead > tr > th {
 
-    If the result of querying by a constructor is different from the result
+    If the result of querying by a media type is different from the result
     of querying by an element, authors should take the latter one as a correct
```
I suggest changing "a correct" to "the correct"

chrisn reviewed (changes requested):
Thanks @alastor0325. I have just made a few more comments
